### ISSUE TITLE
Main decoder error: Playback freezes after some minutes

### ISSUE THREAD

**Issue by lsiem** [caprica/vlcj#1024]
I have an application that should permanently play only a list of media.
This includes MP4 and JPG files.
The media are retrieved locally.
After the player has been running for about 20 minutes without problems, it will freeze at some point.
I get three messages every time the player proceeds to the next media:

`mmal_codec decoder: VCSM init succeeded: CMA`
which I believe is nothing bad

`main decoder error: buffer deadlock prevented`
I don't really understand this error. I cannot find anything useful on the internet :/

`vlcpulse audio output error: PulseAudio server connection failure: Connection refused`
which is ok since I don't have any audio to play.

Here is the important part of my code:

```
    VLC_ARGS.add("--demux=avformat");
    MediaPlayerFactory mediaPlayerFactory = new MediaPlayerFactory(VLC_ARGS);
    Canvas canvas = new Canvas();
    canvas.setBackground(Color.black);
    VideoSurface videoSurface = mediaPlayerFactory.videoSurfaces().newVideoSurface(canvas);
    EmbeddedMediaPlayer mediaPlayer = mediaPlayerFactory.mediaPlayers().newEmbeddedMediaPlayer();
    mediaPlayer.videoSurface().set(videoSurface);

    mediaListPlayer = mediaPlayerFactory.mediaPlayers().newMediaListPlayer();
    mediaListPlayer.mediaPlayer().setMediaPlayer(mediaPlayer);

    JPanel panel = new JPanel();
    panel.setBackground(Color.black);
    panel.setLayout(new BorderLayout());
    panel.add(canvas, BorderLayout.CENTER);

    JFrame frame = new JFrame("VidamaPlayer");
    frame.setContentPane(panel);
    frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
    frame.setSize((int) Toolkit.getDefaultToolkit().getScreenSize().getWidth(),
        (int) Toolkit.getDefaultToolkit().getScreenSize().getHeight());
    frame.setUndecorated(true);
    frame.setVisible(true);
    mediaPlayer.fullScreen().strategy(new XFullScreenStrategy(frame));

    frame.addKeyListener(new KeyListener() {
      @Override
      public void keyTyped(KeyEvent e) {
        // empty
      }

      @Override
      public void keyPressed(KeyEvent e) {
        if (e.getKeyCode() == KeyEvent.VK_Q) {
          frame.dispose();
          mediaListPlayer.release();
          mediaPlayer.release();
          System.exit(1);
        }
      }

      @Override
      public void keyReleased(KeyEvent e) {
        // empty
      }
    });

    try {
      playerLoop();
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
  }

  public void playerLoop() throws InterruptedException {
    String imgDuration = "image-duration=5";

    MediaList playlist = new MediaPlayerFactory().media().newMediaList();

```
... some other nonrelated stuff...

```
    MediaListRef mediaListRef = playlist.newMediaListRef();
    try {
      mediaListPlayer.list().setMediaList(mediaListRef);
    } finally {
      mediaListRef.release();
    }

    mediaListPlayer.controls().setMode(PlaybackMode.LOOP);
    mediaListPlayer.controls().play();
    Thread.currentThread().join();

```

Here are some details about my running system:

Raspberry Pi 4b 4GB
Raspberry Pi OS with Linux 5.4.72
VLCJ Version: 4.5.2
VLC Version: 3.0.11

Thank you in advance!

**Comment by caprica** [comment]
Generally, I know for sure vlcj can play media for hours without any problems.

The most common mistake is to not keep hard references to your media player, but that's impossible for me to see unless you post a complete representative class or classes, rather than pulling out these disjointed fragments.

Your buffer deadlock and pulse audio errors are not to be ignored here and are almost certainly symptoms of the problem.

**Comment by caprica** [comment]
For example, the code you posted, you create this:
```
EmbeddedMediaPlayer mediaPlayer = mediaPlayerFactory.mediaPlayers().newEmbeddedMediaPlayer();
```
It looks like the method that creates that media player just exits normally and when it does the mediaPlayer object goes out of scope and becomes eligible for garbage collection. Depending on the garbage collector this can happen immediately or some indeterminate time later. When that object becomes garbage collected, generally your application will fatally crash.

If this IS your problem, then you must keep a reference to that media player (and depending on how your code is structured you may need to keep a reference to the object that has the reference to your media player.

The most common approach is to move the automatic/heap reference to be class field instead - but like I said, you must also ensure that the class itself is not eligible for garbage collection.

This is why I asked you to post full code rather than fragments, because it's hard to see if this IS your problem given what you posted.

**Comment by lsiem** [comment]
First of all thanks for the detailed and informative answer!
The problem may very well be the garbage collection. I did not even think of that. Thanks for the hint!
Here is my complete class:


```
public class VidamaPlayer {

  private static VidamaPlayer vidamaPlayer;
  private MediaListPlayer mediaListPlayer;
  private EmbeddedMediaPlayer mediaPlayer;
  private static final String MEDIA_PATH_UNIX = System.getProperty("user.home") + "/media/";
  private static final String MEDIA_PATH_WINDOWS = System.getProperty("user.home") + "\\media\\";
  private static final List<String> VLC_ARGS = new ArrayList<>();

  public static void main(String[] args) {
    vidamaPlayer = new VidamaPlayer();
    vidamaPlayer.initialize();
  }

  public void initialize() {
    VLC_ARGS.add("--demux=avformat");
    MediaPlayerFactory mediaPlayerFactory = new MediaPlayerFactory(VLC_ARGS);
    Canvas canvas = new Canvas();
    canvas.setBackground(Color.black);
    VideoSurface videoSurface = mediaPlayerFactory.videoSurfaces().newVideoSurface(canvas);
    mediaPlayer = mediaPlayerFactory.mediaPlayers().newEmbeddedMediaPlayer();
    mediaPlayer.videoSurface().set(videoSurface);

    mediaListPlayer = mediaPlayerFactory.mediaPlayers().newMediaListPlayer();
    mediaListPlayer.mediaPlayer().setMediaPlayer(mediaPlayer);

    JPanel panel = new JPanel();
    panel.setBackground(Color.black);
    panel.setLayout(new BorderLayout());
    panel.add(canvas, BorderLayout.CENTER);

    JFrame frame = new JFrame("VidamaPlayer");
    frame.setContentPane(panel);
    frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
    frame.setSize((int) Toolkit.getDefaultToolkit().getScreenSize().getWidth(),
        (int) Toolkit.getDefaultToolkit().getScreenSize().getHeight());
    frame.setUndecorated(true);
    frame.setVisible(true);
    mediaPlayer.fullScreen().strategy(new XFullScreenStrategy(frame));

    frame.addKeyListener(new KeyListener() {
      @Override
      public void keyTyped(KeyEvent e) {
        // empty
      }

      @Override
      public void keyPressed(KeyEvent e) {
        if (e.getKeyCode() == KeyEvent.VK_Q) {
          frame.dispose();
          mediaListPlayer.release();
          mediaPlayer.release();
          System.exit(1);
        }
      }

      @Override
      public void keyReleased(KeyEvent e) {
        // empty
      }
    });

    try {
      playerLoop();
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
  }

  public void playerLoop() throws InterruptedException {
    String imgDuration = "image-duration=5";

    MediaList playlist = new MediaPlayerFactory().media().newMediaList();

    URL url = null;
    String urlPreset = "https://vidama.lsiem.de/media/new/";
    try {
      url = new URL(urlPreset + "playlist.json");
    } catch (MalformedURLException e) {
      e.printStackTrace();
    }
    assert url != null;
    List<String> fileNames = new ArrayList<>();
    try (Scanner scanner = new Scanner(url.openStream(), StandardCharsets.UTF_8)) {
      String json = scanner.useDelimiter("\\A").next();
      JSONObject jsonObject = new JSONObject(json);
      JSONArray jsonArray = jsonObject.getJSONArray("media");
      for (int i = 0; i < jsonArray.length(); i++) {
        fileNames.add(jsonArray.getString(i));
      }
    } catch (IOException e) {
      e.printStackTrace();
    }

    if (System.getProperty("os.name").startsWith("Windows")) {
      if (!Files.exists(Path.of(MEDIA_PATH_WINDOWS))) {
        try {
          Files.createDirectory(Path.of(MEDIA_PATH_WINDOWS));
        } catch (IOException e) {
          e.printStackTrace();
        }
      }
      for (String fileName : fileNames) {
        downloadMedia(urlPreset + fileName, fileName, true);
        if (fileName.endsWith("jpg")) {
          playlist.media().add(MEDIA_PATH_WINDOWS + fileName, imgDuration);
        } else {
          playlist.media().add(MEDIA_PATH_WINDOWS + fileName);
        }
      }
    } else {
      if (!Files.exists(Path.of(MEDIA_PATH_UNIX))) {
        try {
          Files.createDirectory(Path.of(MEDIA_PATH_UNIX));
        } catch (IOException e) {
          e.printStackTrace();
        }
      }
      for (String fileName : fileNames) {
        downloadMedia(urlPreset + fileName, fileName, false);
        if (fileName.endsWith("jpg")) {
          playlist.media().add(MEDIA_PATH_UNIX + fileName, imgDuration);
        } else {
          playlist.media().add(MEDIA_PATH_UNIX + fileName);
        }
      }
    }

    MediaListRef mediaListRef = playlist.newMediaListRef();
    try {
      mediaListPlayer.list().setMediaList(mediaListRef);
    } finally {
      mediaListRef.release();
    }

    mediaListPlayer.controls().setMode(PlaybackMode.LOOP);
    mediaListPlayer.controls().play();
    Thread.currentThread().join();
  }

  private void downloadMedia(String url, String name, boolean windows) {
    try (InputStream inputStream = new URL(url).openStream()) {
      if (windows) {
        if (Files.exists(Paths.get(MEDIA_PATH_WINDOWS + name))) {
          return;
        }
        Files.copy(inputStream, Path.of(MEDIA_PATH_WINDOWS, name),
            StandardCopyOption.REPLACE_EXISTING);
      } else {
        if (Files.exists(Paths.get(MEDIA_PATH_UNIX + name))) {
          return;
        }
        Files.copy(inputStream, Path.of(MEDIA_PATH_UNIX, name),
            StandardCopyOption.REPLACE_EXISTING);
      }
    } catch (IOException e) {
      Logger.error("Fehler beim herunterladen von Medien: " + e);
    }
  }
}
```
I have already extracted some instances from the local scope.
I experienced another error message while testing:
`cache_read stream error: cannot pre fill buffer`

Thanks again for the help. Should I rather do a stackoverflow thread, since this doesn't seem to be a bug of VLCJ, but a problem on my part?

**Comment by caprica** [comment]
You can often ignore those "errors" from VLC to be honest.

With the code you posted above, do you still experience the same problem that you originally reported?

**Comment by lsiem** [comment]
Yes, the problem unfortunately still occurs.
Once the error message:
`cache_read stream error: cannot pre fill buffer`
comes, the player freezes.

**Comment by caprica** [comment]
Can't see anything obviously wrong.

Try creating your factory with "-vvv" and then inspect the verbose logs when it fails.

**Comment by lsiem** [comment]
I saved the logs as a file and looked at them. 
The following messages appear when freezing:

```
main debug: killing decoder fourcc `h264'
main debug: removing module "mmal_codec"
main debug: saving a free vout
main debug: reusing provided vout
main debug: removing module "mp4"
mp4 debug: freeing all memory
main debug: Program doesn't contain anymore ES
main debug: removing module "record"
main debug: removing module "cache_read"
main debug: removing module "filesystem"
main debug: Creating an input for '1080x1920_EWE_final.jpg'
main debug: using timeshift granularity of 50 MiB
main debug: using default timeshift path
main debug: `file:///home/pi/media/1080x1920_EWE_final.jpg' gives access `file' demux `any' path `/home/pi/media/1080x1920_EWE_final.jpg'
main debug: creating demux: access='file' demux='any' location='/home/pi/media/1080x1920_EWE_final.jpg' file='/home/pi/media/1080x1920_EWE_final.jpg'
main debug: looking for access_demux module matching "file": 17 candidates
main debug: no access_demux modules matched
main debug: creating access: file:///home/pi/media/1080x1920_EWE_final.jpg
main debug:  (path: /home/pi/media/1080x1920_EWE_final.jpg)
main debug: looking for access module matching "file": 28 candidates
main debug: using access module "filesystem"
main debug: looking for stream_filter module matching "prefetch,cache_read": 26 candidates
cache_read debug: Using stream method for AStream*
cache_read debug: starting pre-buffering
cache_read debug: pre-buffering done 0 bytes in 0s - 0 KiB/s
cache_read error: cannot pre fill buffer
main debug: no stream_filter modules matched
main warning: cannot insert stream filter prefetch,cache_read
main debug: looking for stream_filter module matching "any": 26 candidates
playlist debug: not enough data
lua debug: Trying Lua scripts in /home/pi/.local/share/vlc/lua/playlist
lua debug: Trying Lua scripts in /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/anevia_streams.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/anevia_xml.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/appletrailers.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/bbc_co_uk.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/cue.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/dailymotion.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/jamendo.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/koreus.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/liveleak.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/newgrounds.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/rockbox_fm_presets.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/soundcloud.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/twitch.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/vimeo.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/vocaroo.luac
lua debug: Trying Lua playlist script /usr/lib/arm-linux-gnueabihf/vlc/lua/playlist/youtube.luac
lua debug: Trying Lua scripts in /usr/share/vlc/lua/playlist
main debug: no stream_filter modules matched
main debug: looking for stream_directory module matching "any": 1 candidates
main debug: no stream_directory modules matched
main debug: attachment of directory-extractor failed for file:///home/pi/media/1080x1920_EWE_final.jpg
main debug: looking for stream_filter module matching "record": 26 candidates
main debug: using stream_filter module "record"
main debug: creating demux: access='file' demux='any' location='/home/pi/media/1080x1920_EWE_final.jpg' file='/home/pi/media/1080x1920_EWE_final.jpg'
main debug: looking for demux module matching "any": 55 candidates
es debug: cannot peek
es debug: cannot peek
es debug: cannot peek
es debug: cannot peek
es debug: cannot peek
mod debug: MOD validation failed (ext=jpg)
ps debug: cannot peek
mjpeg warning: no more data
mjpeg error: cannot peek
mpgv debug: cannot peek
avformat warning: cannot peek
vobsub debug: could not read vobsub IDX file
ps debug: cannot peek
main debug: no demux modules matched
main debug: removing module "record"
main debug: removing module "filesystem"
main debug: destroying useless vout
main debug: removing module "mmal_codec"
main debug: Filter 0x5f723c00 removed from chain
main debug: removing module "mmal_xsplitter"
mmal_xsplitter debug: <<< CloseMmalX11
main debug: removing module "gles2"
main debug: removing module "mmal_converter"
main debug: removing module "egl_x11"
main debug: removing module "mmal_vout"
mmal_xsplitter debug: >>> CloseMmalX11
main debug: removing module "xdg"
main debug: removing module "freetype"
main debug: removing module "yuvp"
main debug: removing module "swscale"
```

`cannot peek` Could this possibly be the cause?

**Comment by caprica** [comment]
Really don't know, it's very doubtful there's a vlcj problem here.

Are you only playing static images? And they're all on your local filesystem?

Why is it looking for stream_filter "record", so many questions...

The image is definitely not corrupted in some way is it?

**Comment by lsiem** [comment]
I play pictures and videos from the local file system. 
These are downloaded from a website based on the entries in a JSON file.
The media is played back permanently as a loop. The image is initially played back without any problems. Only after the 5th or 6th loop, the player will freeze.

**Comment by caprica** [comment]
But are you sure the files are downloading correctly?

**Comment by caprica** [comment]
I don't really have any suggestions here.
